### DISCUSSION TITLE
Code not compiling when ganged axis is turned on

### DISCUSSION THREAD

**Discussion by rvalotta** [terjeio/grblHAL#343] [Q&A]
Just updated my src and tried to recompile it. I'm getting the following

```
In file included from Src\driver.c:29:0:
Src\driver.c: In function 'stepperSetDirOutputs':
Src\driver.c:665:145: error: request for member 'y' in something not a structure or union
     DIGITAL_OUT(Y2_DIRECTION_PORT, Y2_DIRECTION_PIN, (dir_outbits.y ^ settings.steppers.dir_invert.y) ^ settings.steppers.ganged_dir_invert.mask.y);
                                                                                                                                                 ^
Inc/driver.h:46:71: note: in definition of macro 'DIGITAL_OUT'
 #define DIGITAL_OUT(port, pin, on) { BITBAND_PERI((port)->ODR, pin) = on; }
```
The code in question is the following (driver.c line 665)
```
  #ifdef Y2_DIRECTION_PIN
    DIGITAL_OUT(Y2_DIRECTION_PORT, Y2_DIRECTION_PIN, (dir_outbits.y ^ settings.steppers.dir_invert.y) ^ settings.steppers.ganged_dir_invert.mask.y);
  #endif
```

Tracing it back there is no .y as a member of .mask.  Should that be removed?

**Comment by rvalotta** [comment]
I get more and more preplexed as i compile this. Just to be clear this is for the STM32F4XX driver. I'm using the blackpill_map.h.  Why would we disable the proble pin if y axis is Ganged?

```
// Define probe switch input pin.
#if !N_AUTO_SQUARED
#define PROBE_PORT              GPIOB
#define PROBE_PIN               10
#endif
```

**Reply by rvalotta** [reply]
And for what ever reason the VFD spindle control isn't working... Guess we go back a few revs and figure out where it broke. That may be a few days from now.

**Reply by rvalotta** [reply]
Re-inserted my previous STM32F4XX and i'm back up and running  for reference here is the build info 

[VER:1.1f.20211029:]
[OPT:VNMH0SL2,35,1024,3,0]
[NEWOPT:ENUMS,RT+,HOME,TC,SED]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[DRIVER:STM32F411]
[DRIVER VERSION:211101]
[BOARD:BlackPill]
[PLUGIN:MODBUS v0.07]
[PLUGIN:HUANYANG VFD v0.03]

**Reply by terjeio** [reply]
> Tracing it back there is no .y as a member of .mask. Should that be removed?

It is a bug, `.mask` has to be removed. Sorry for that, I'll fix that in the next commit.

`settings.steppers.ganged_dir_invert.y`

>  Why would we disable the proble pin if y axis is Ganged?

A ganged axis does not need a second limit pin, an auto squared one does. The blackpill has a limited number of pins available (that is interrupt capable) and I guess that is the reason. Be aware that a given pin number regardless of which port it belongs to can only have one interrupt assigned. Eg. B12 and C12 cannot both be used for control or limit inputs. This is a limitation of the STM32 interrupt peripheral.

**Reply by rvalotta** [reply]
Thanks I learnt that lesson the hard way a few days ago. Was trying to remap things to make for a better pcb layout and well it took some reading and googling to figure that out.   

As for limited pins I'm not really feeling it.... right now I have pins mapped out for 4 axis (1 is auto squared), Reset, Hold, Feed, Door Safety, Probe, Step En/Dis, PWM for diode, RS485 for VFD, coolant flood and coolant myst and 5 aux pins (3 out 2 in). Granted not all battle tested yet but they seem good so far.

Any suggestions on where to start with regards to the VFD spindle not working anymore? Comes back with a Alarm 14 right away.

**Reply by dresco** [reply]
> Any suggestions on where to start with regards to the VFD spindle not working anymore? Comes back with a Alarm 14 right away.

fwiw, have just pulled & built the latest code here, and is working fine in my setup (have a spare vfd with my dev board). I've not enabled the spindle at speed check though?

Are you running the default baud rate of 19,200? My dev setup is 38,400 - and that often catches me out when I forget to change the baud rate after flashing..

**Reply by rvalotta** [reply]
Yup Default 19200, checked my working STM32 (I keep two and rotate them between revisions).  

Spindle at speed is not turned on

Compiling using Visual Studio and Platform IO
With the following enabled

-D BOARD_BLACKPILL=
-D USB_SERIAL_CDC=1
-D VFD_ENABLE=1
-D DUAL_SPINDLE=1
-D MODBUS_ENABLE=1
-D Y_AUTO_SQUARE=1

My tweaked blackpill_map file can be seen here https://pastebin.com/m80DVWqm
My Settings: https://pastebin.com/MWdcyDD3

**Reply by rvalotta** [reply]
Got this one figured out.    Seems there is some come left from testing modbus that was using UART2 for debugging and those were the pins i was using for my VFD.  Seems there are now dedicated pins for the modbus. I will be testing this new setup in the next few days.
